### PR TITLE
Use parse_url kernel for PROTOCOL parsing

### PR DESCRIPTION
**Pull request by thirtiseven**
Contributes to #8963 

This PR add plugin support for PROTOCOL parsing in `parse_url`.

Related JNI PR: https://github.com/NVIDIA/spark-rapids-jni/pull/1502

Most of the code is reused from the old PR #8761 

**Performance test:**
```
spark.time(df.selectExpr("COUNT(parse_url(url, 'PROTOCOL')) as pr1", "COUNT(parse_url(url, 'PROTOCOL')) as pr2", "COUNT(parse_url(url, 'PROTOCOL')) as pr3", "COUNT(parse_url(url, 'PROTOCOL')) as pr4", "COUNT(parse_url(url, 'PROTOCOL')) as pr5", "COUNT(parse_url(url, 'PROTOCOL')) as pr6", "COUNT(parse_url(url, 'PROTOCOL')) as pr7", "COUNT(parse_url(url, 'PROTOCOL')) as pr8", "COUNT(parse_url(url, 'PROTOCOL')) as pr9", "COUNT(parse_url(url, 'PROTOCOL')) as pr0").show())
```
against 651191 urls' dataset, repeated 100 times.

CPU: 10752 ms
GPU: 1032 ms

It's about 10x faster

### CODE REVIEW

File: docs/compatibility.md
Comment (jlowe) [inline docs/compatibility.md:463]: If these are still the case, shouldn't we fix the parsing kernel then?  The whole point of having a custom parsing kernel is to match Spark's behavior.  cc: @hyperbolic2346

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3241]: I don't see this checking for whether this is just for protocol parsing.  There should be fallback tests for all the cases that we don't support (i.e.: all the other parse part directive literals, parse directive as column rather than literal, etc.)

File: tests/src/test/scala/com/nvidia/spark/rapids/UrlFunctionsSuite.scala
Comment (jlowe) [inline tests/src/test/scala/com/nvidia/spark/rapids/UrlFunctionsSuite.scala:21]: Should this just be moved into the pyspark integration test?  I don't see anything this test is doing that cannot be done in the integration test, and the latter is preferable since it executes in parallel.

File: docs/compatibility.md
Comment (hyperbolic2346) [reply docs/compatibility.md:463]: Completely agree and we will.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala
Comment (thirtiseven) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3240]: done.

File: tests/src/test/scala/com/nvidia/spark/rapids/UrlFunctionsSuite.scala
Comment (thirtiseven) [reply tests/src/test/scala/com/nvidia/spark/rapids/UrlFunctionsSuite.scala]: Ok, moved to integration test.

File: integration_tests/src/main/python/url_test.py
Comment (jlowe) [inline integration_tests/src/main/python/url_test.py:226]: This seems like a test of Spark itself rather than the plugin?  IIUC this error is going to be thrown not because the plugin throws it but during the query analysis which occurs before the plugin runs.

File: integration_tests/src/main/python/url_test.py
Comment (jlowe) [inline integration_tests/src/main/python/url_test.py:161]: Nit: Parameterize this test on the data generation rather than duplicate it

File: integration_tests/src/main/python/url_test.py
Comment (jlowe) [inline integration_tests/src/main/python/url_test.py:197]: all of these should be refactored to a single test parameterized on the literal rather than duplicated.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3235]: This should have a `withPsNote` denoting that it only supports PROTOCOL so that limitation shows up in the documentation.  Otherwise the documentation implies we fully support this as long as a literal is supplied.

File: integration_tests/src/main/python/url_test.py
Comment (jlowe) [inline integration_tests/src/main/python/url_test.py:38]: The failing cases need to be uncommented and passing before this should be merged.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3244]: Thais is not guaranteed to work. You said it was a `lit`, but `lit`is more lenient that just a Literal. please use `extractStringLit` instead. 

https://github.com/NVIDIA/spark-rapids/blob/ed3a1f3e879abd7b0574cbeb990eaa8f421ea47d/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala#L591

i.e.

```
extractStringLit(childExprs(1)).map(_.toUpperCase) match {
  case Some("PROTOCOL") =>
  case Some(other) =>
    willNotWorkOnGpu(s"Part to extract $other is not supported on GPU")
  case None =>
    // Should never get here, but just in case
    willNotWorkOnGpu("GPU only supports a literal for the part to extract")
}
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala:39]: Should these be public and the GpuOverrides uses them instead of hard coding values there?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala:46]: If we don't support it should we even have it here? Should there be an assertion if it is set?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala:68]: Can we remove this as it is never called?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala:98]: If the URL can never be a scalar then don't call columnarEvalAny on it. Just call `columnarEval` which will return a `GpuColumnVector` and use that without the need for the special match.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala:112]: Same here just call `columnarEval` to get back a vector, guaranteed, and then don't include it in the match.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala:99]: Can we rename val0, val1, and val2 to something a little more descriptive?

File: integration_tests/src/main/python/url_test.py
Comment (thirtiseven) [reply integration_tests/src/main/python/url_test.py]: Yes you are right! I removed this case and related unnecessary error handling code.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3244]: nit: For me I would rather that we just reach into a and pull out the expression we care about.

```
a.children(1)
```

but this is really minor.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala:78]: nit: no need to match URLs here or use the IfAllowed version of withResource.

```
withResource(url.columnarEval(batch)) { urls =>
  withResourceIfAllowed(partToExtract.columnarEvalAny(batch)) { parts =>
    parts match {
      case partScalar: GpuScalar =>
        GpuColumnVector.from(doColumnar(urls, partScalar), dataType)
      case _ => ...
      }
    }
  }
}
```

This is also really minor, but it might clean things up a little. The same comment goes for the 3 arg version.

**Review by revans2 — COMMENTED** [review]
Just two nits that I don't really care about. The only reason I am not approving it is because we want to wait until all of the URLs for the tests can pass.

**Comment by thirtiseven** [comment]
Depends on https://github.com/NVIDIA/spark-rapids-jni/pull/1554

File: docs/additional-functionality/advanced_configs.md
Comment (hyperbolic2346) [inline docs/additional-functionality/advanced_configs.md:303]: should this be URI or URL?

File: docs/compatibility.md
Comment (hyperbolic2346) [reply docs/compatibility.md:479]: This should be updated to the current implementation, which I believe is 100% identical.

**Review by hyperbolic2346 — CHANGES_REQUESTED** [review]
I think the tests can be enabled now. Question about URL vs URI, but I'm fine either way.

File: docs/additional-functionality/advanced_configs.md
Comment (thirtiseven) [reply docs/additional-functionality/advanced_configs.md:303]: We need to implement a URI parser in jni because Java uses it's URI library to parse urls, but this function is intended to support url parsing only.

**Comment by thirtiseven** [comment]
Thanks all for the review! I think all comments have been addressed and it's ready to merge once jni PR is merged.

File: integration_tests/src/main/python/url_test.py
Comment (thirtiseven) [reply integration_tests/src/main/python/url_test.py]: done.

File: integration_tests/src/main/python/url_test.py
Comment (thirtiseven) [reply integration_tests/src/main/python/url_test.py]: done.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala]: done.

**Comment by sameerz** [comment]
Please retarget to 24.02.

**Comment by thirtiseven** [comment]
Verified locally that all protocol tests passed after https://github.com/NVIDIA/spark-rapids-jni/pull/1554 merged, both for IT and the kaggle dataset.

**Comment by thirtiseven** [comment]
build

**Comment by thirtiseven** [comment]
build

**Comment by thirtiseven** [comment]
Hi @revans2 @jlowe Could you please take another look at this?

**Review by revans2 — APPROVED** [review]

**Comment by revans2** [comment]
@hyperbolic2346 cold you also take a look as your change request is blocking this from going in.

@thirtiseven I think @jlowe is out on vacation so you probably should not wait for him to come back.

On a side note, this is crazy fast!

I ran 
```
spark.time(spark.range(0, 1000000000L, 1, 48).selectExpr("CONCAT('htpp://', CAST(id as STRING)) as u").selectExpr("parse_url(u, 'PROTOCOL') as p").selectExpr("MAX(p)", "MIN(p)").show())
```

both on the GPU and the CPU and the GPU ended up being 452x faster than 12 CPU cores. This is a crappy test that is not real world at all, but still. That is really fast.

**Review by hyperbolic2346 — APPROVED** [review]
I'm fine with this, but I am not a java/scala reviewers

**Comment by revans2** [comment]
> I'm fine with this, but I am not a java/scala reviewer.

I get that, but I just don't like to use my admin merge powers to push a red merge button unless I have to.

**Pull request merged**
